tests/acceptance/conftest.py

```diff
@@ -0,0 +1,106 @@
+from contextlib import contextmanager
+
+import pytest
+
+from databuilder.backends import TPPBackend
+from databuilder.dsl import Cohort
+from tests.lib.tpp_schema import (
+    apcs,
+    ctv3_event,
+    negative_test,
+    organisation,
+    patient,
+    patient_address,
+    positive_test,
+    registration,
+    snomed_event,
+)
+from tests.lib.util import extract
+
+
+@pytest.fixture(scope="module")
+def populated_database(database):
+    setup_data(database)
+    yield database
+
+
+@pytest.fixture
+def cohort(populated_database):
+    class AssertingCohort(Cohort):
+        def assert_results(self, **results):
+            dataset = extract(self, TPPBackend, populated_database)[0]
+            for variable, expected_value in results.items():
+                assert dataset[variable] == expected_value
+
+    yield AssertingCohort()
+
+
+@pytest.fixture
+def subtest(subtests):
+    @contextmanager
+    def helper(name, missing_feature=""):
+        with subtests.test(msg=name):
+            try:
+                yield
+            except Exception:
+                if missing_feature:
+                    pytest.xfail(f"DSL feature '{missing_feature}' not implemented")
+                raise
+            assert (
+                not missing_feature
+            ), f"Unexpected success, maybe feature '{missing_feature}' has been implemented"
+
+    yield helper
+
+
+def setup_data(database):
+    database.setup(
+        organisation(organisation_id=1, region="South"),
+        patient(
```

Comment:
It might make the tests clearer if we allow this to vary by study rather than having a single data fixture for all of these tests.

Reply:
It would, especially as the number of tests increases. We could:

* Move `populated_database` (the fixture) and `setup_data` back to `test_long_covid_dsl`
* Merge `setup_data` into `populated_database`
* Write `test_sro_measures_dsl.populated_database` (i.e. we would have `test_long_covid_dsl.populated_database` and `test_sro_measures_dsl.populated_database`)

What was your rationale for separating `populated_database` and `setup_data`?

Alternatively, we could:

* Keep `populated_database` where it is, but pass it the [`request`][2] object
* Move `setup_data` back to `test_long_covid_dsl`
* Write `test_sro_measures_dsl.setup_data`
* Introspect `request`, for the module-specific `setup_data`

This keeps `populated_database` and `setup_data` separate, but might seem 🪄.

[1]: https://ebmdatalab.slack.com/archives/C02LEF2GC75/p1642700504246100
[2]: https://docs.pytest.org/en/stable/fixture.html#fixtures-can-introspect-the-requesting-test-context

Reply:
> What was your rationale for separating populated_database and setup_data?

A minor point of taste. Separating the concerns of "define a fixture that populates the database with the data needed for this test" and "define the data needed for this test".

Reply:
I don't think it's quite a simple as you suggest (at least in your first option) because the populated database is a dependency (as currently written) of the `cohort` fixture.

Since the `populated_database` fixture is so simple, maybe it's okay to have it duplicated between tests. So you could explicitly pull that into every test.

```py
## conftest.py

@pytest.fixture
def cohort(database):
    # ...
    
## test_long_covid_dsl.py

def setup_data(database):
    # ...

@pytest.fixture
def populated_database(database):
    # ...
    
def test_long_covid_study(cohort, subtest, populated_database):
    # ...
    
## test_sro_measures_dsl.py

def setup_data(database):
    # ...

@pytest.fixture
def populated_database(database):
    # ...
    
def test_sro_measures_study(cohort, subtest, populated_database):
    # ...
```

Maybe?

Maybe that's overkill and this is simpler:

```py
def test_sro_measures_study(cohort, subtest, database):
    setup_data(database)
    
    # ...
```

etc

Since there's only one test in each of these files, we don't really gain anything by pulling the database population out into a fixture, do we?

Reply:
I've used the `request` object to introspect the module, as this handles the dependency between `cohort` and `populated_database`. As a consequence, a test that uses the `cohort` fixture will result in a call to the `setup_data` function from within that test's module. Just because we can do it this way, though, doesn't mean we should: how do you feel about the additional 🪄?

Something I'd like to check: regardless of approach, we don't truncate the database between calls. Should we?

Reply:
Ah, but we do truncate the database between calls!

https://github.com/opensafely-core/databuilder/blob/c605cd2a58d2669d4da6010951269c1f058b7e12/tests/lib/databases.py#L97-L98

Reply:
I think that's a bit too much magic. Even someone super-familiar with pytest won't have any reason to expect that.

Reply:
If we took the simpler path, then would you pass `database` as an argument to `AssertingCohort.assert_results`?

Reply:
Oh, hold on, we don't have to, do we? We could just make the `cohort` fixture depend on the `database` fixture and remember that the `setup_data` function must be called before `AssertingCohort.assert_results` is called.

Reply:
Yes, that's right.

Reply:
I've removed the 🪄 and taken the simpler path. Consequently, we no longer need the `populated_database` fixture, in either its original or modified form.